math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -276,163 +1214,179 @@ auto hpt = d.Define("pt", "sqrt(pxs * pxs + pys * pys)[E>200]")
-   using value_type = typename Impl_t::value_type;
-   using size_type = typename Impl_t::size_type;
-   using difference_type = typename Impl_t::difference_type;
-   using reference = typename Impl_t::reference;
-   using const_reference = typename Impl_t::const_reference;
-   using pointer = typename Impl_t::pointer;
-   using const_pointer = typename Impl_t::const_pointer;
-   // The data_t and const_data_t types are chosen to be void in case T is a bool.
-   // This way we can as elegantly as in the STL return void upon calling the data() method.
-   using data_t = typename std::conditional<IsVecBool, void, typename Impl_t::pointer>::type;
-   using const_data_t = typename std::conditional<IsVecBool, void, typename Impl_t::const_pointer>::type;
-   using iterator = typename Impl_t::iterator;
-   using const_iterator = typename Impl_t::const_iterator;
-   using reverse_iterator = typename Impl_t::reverse_iterator;
-   using const_reverse_iterator = typename Impl_t::const_reverse_iterator;
-
-private:
-   Impl_t fData;
+   static constexpr unsigned N = 8;
+   RVec() : Internal::VecOps::SmallVectorImpl<T>(N) {}
 
-public:
-   // constructors
-   RVec() {}
+   ~RVec()
+   {
+      if (this->Owns) {
+         // Destroy the constructed elements in the vector.
+         this->destroy_range(this->begin(), this->end());
+      }
+   }
 
-   explicit RVec(size_type count) : fData(count) {}
+   explicit RVec(size_t Size, const T &Value = T()) : Internal::VecOps::SmallVectorImpl<T>(N)
+   {
+      this->assign(Size, Value);
+   }
 
-   RVec(size_type count, const T &value) : fData(count, value) {}
+   template <typename ItTy,
+             typename = typename std::enable_if<std::is_convertible<
+                typename std::iterator_traits<ItTy>::iterator_category, std::input_iterator_tag>::value>::type>
+   RVec(ItTy S, ItTy E) : Internal::VecOps::SmallVectorImpl<T>(N)
+   {
+      this->append(S, E);
+   }
 
-   RVec(const RVec<T> &v) : fData(v.fData) {}
+   RVec(std::initializer_list<T> IL) : Internal::VecOps::SmallVectorImpl<T>(N) { this->assign(IL); }
 
-   RVec(RVec<T> &&v) : fData(std::move(v.fData)) {}
+   RVec(const RVec &RHS) : Internal::VecOps::SmallVectorImpl<T>(N)
+   {
+      if (!RHS.empty())
+         Internal::VecOps::SmallVectorImpl<T>::operator=(RHS);
+   }
 
-   RVec(const std::vector<T> &v) : fData(v.cbegin(), v.cend()) {}
+   RVec &operator=(const RVec &RHS)
+   {
+      Internal::VecOps::SmallVectorImpl<T>::operator=(RHS);
+      return *this;
+   }
 
-   RVec(pointer p, size_type n) : fData(n, T(), ROOT::Detail::VecOps::RAdoptAllocator<T>(p)) {}
+   RVec(RVec &&RHS) : Internal::VecOps::SmallVectorImpl<T>(N)
+   {
+      if (!RHS.empty())
+         Internal::VecOps::SmallVectorImpl<T>::operator=(::std::move(RHS));
+   }
 
-   template <class InputIt>
-   RVec(InputIt first, InputIt last) : fData(first, last) {}
+   RVec(Internal::VecOps::SmallVectorImpl<T> &&RHS) : Internal::VecOps::SmallVectorImpl<T>(N)
```

Comment:
What other class might derive from `SmallVectorImpl<T>`? I get this in `llvm::SmallVector`'s case where the small buffer optimization size is a template parameter. But here?

Reply:
I removed this